File: Problems/3_Longest_Substring_Without_Repeating_Characters.cpp
#include<iostream>
#include<string>

using namespace std;

// using set makes it faster
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        if (!s.length()) return 0;
        unordered_set<char> seen;
        int l = 0;
        int maxLen = 1;
        for (int r = 0; r < s.length(); ++r) {
            while (seen.count(s[r])) {
                seen.erase(s[l]);
                ++l;
            }
            seen.insert(s[r]);
            maxLen = max(maxLen, r - l + 1);
        }
        return maxLen;
    }
};

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int len = s.size();
        int lastSeen[95];
        for (int i = 0; i < 95; ++i)
            lastSeen[i] = -1;
        int left = 0;
        int maxLen = 0;
        for (int right = 0; right < s.length(); ++right){
            left = max(left, lastSeen[s[right]-32]+1);
            maxLen = max(maxLen, right-left+1);
            lastSeen[s[right]-32] = right;
        }
        return maxLen;
    }
};

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int max_len = 0;
        string sub_string;
        for(auto c: s){
            while(sub_string.find(c) != string::npos){
                sub_string.erase(0, 1);
            }
            sub_string += c;
            max_len = max<int>(max_len, sub_string.length());
        }
        return max_len;
    }
};

int main(){
    Solution sol;
    string s =  "abcabcbb";
    cout <<sol.lengthOfLongestSubstring(s)<<endl;
    return 0;
}

File: Problems/3/3_Longest_Substring_Without_Repeating_Characters.py
# Best Solution using set
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        l = 0
        window = set()
        res = 0
        for r in range(len(s)):
            while s[r] in window:
                window.remove(s[l])
                l += 1
            window.add(s[r])
            res = max(res, r - l + 1)
        return res

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s)<1:
            return len(s)
        left = 0
        maxLen = 1
        lastSeen = [-1]*95
        for right in range(len(s)):
            c = s[right]
            left = max(left, lastSeen[ord(c)-32]+1)
            lastSeen[ord(c)-32] = right
            maxLen = max(maxLen, right-left+1)
        return maxLen


class Solution(object):
    def lengthOfLongestSubstring(self, s):
        sub_str = ''
        max_len = 0
        for w in s:
            while w in sub_str:
                sub_str = sub_str[1:]
            sub_str += w
            max_len = max(max_len, len(sub_str))
        
        return max_len

# sol = Solution()
# print(sol.lengthOfLongestSubstring("loddktdji"))


######### Better Solution (with set)    ####################
# class Solution:
#     def lengthOfLongestSubstring(self, s: str) -> int:
#         max_length = 0
#         used_c = set()
#         while max_length < len(s):
#             for i in range(len(s)):
#                 if s[i] not in used_c:
#                     used_c.add(s[i])
#                 else:
#                     if len(used_c) > max_length:
#                         max_length = len(used_c)
#                     used_c = set()
#                     break
#             start = s.index(s[i]) + 1
#             s = s[start:]
#         return max(max_length ,len(used_c))


#########   Bad Solution (with list)    #################
# class Solution:
#     def lengthOfLongestSubstring(self, s: str) -> int:
#         max_length = 0
#         used_c = []
#         while max_length < len(s):
#             for i in range(len(s)):
#                 if s[i] not in used_c:
#                     used_c.append(s[i])
#                 else:
#                     if len(used_c) > max_length:
#                         max_length = len(used_c)
#                     used_c = []
#                     break
#             start = s.index(s[i]) + 1
#             s = s[start:]
#         return max(max_length ,len(used_c))